### DISCUSSION TITLE
Milvus‘s insert and search

### DISCUSSION THREAD

**Discussion by Yang1032** [milvus-io/milvus#30157] [Q&A and General discussion] [answered]
Hi, I have some questions regarding data insertion and retrieval in Milvus:
1. Milvus supports many indexes, such as IVF, HNSW, ScaNN, etc., which are chosen based on different data scales and scenarios. And, each Segment is mentioned as having an independent index. Could you elaborate on what is meant by this independent index? Shouldn't indexes for other large-scale data like HNSW and ScaNN be a single index for the entire data?
2. Google's ScaNN is based on quantized indexes. Does this mean that the index is reconstructed each time new data is inserted?
3. How does Milvus handle incremental data insertion (with potential index reconstruction) and retrieval requests? Is there a priority to ensure index construction after a user's retrieval, or is there a scheduled approach (e.g., nightly) for index updates?
4. Regarding the newly introduced "Growing index," is it designed to handle data from Growing Segments, i.e., data before index creation for streaming data?
Thank you.

**Comment by xiaofan-luan** [comment, marked as answer]
milvus split data into segments, and all data is built on segments.

There are two types of segments,growing and sealed.
Growing segments handles streaming insert, and we build growing index on top of it .
Sealed segments are immutable, and we build hnsw and faiss index on top of it.

**Reply by Yang1032** [reply]
Thank you for your response. 
Each Segment has an independent index. In other words, for indexes like HNSW (suitable for large-scale data) and IVF (suitable for small-scale data), is the actual data scale the same (the data scale of one segment)? 
If we set the segment size smaller,  they don't offer a significant speed advantage over brute-force retrieval ?

**Reply by xiaofan-luan** [reply]
By default, milvus use segment size of 512MB.
If you have larger node and huge dataset, we recommended to use 2GB or 4GB segment size, which can significantly improve your search performance

**Reply by Yang1032** [reply]
Thank you very much for your patient and clear response, I understand now. 
Also, I have another questions: 
1. If I set the size of the segment to be relatively large, and the Growing Segment is not yet persisted in the buffer, what happens if Milvus experiences a failure and needs to restart. Is there a risk of data loss at this point?
2. hh approximately how many rows of vectors (e.g., 1024 dimensions) can be stored in 512MB?

**Reply by xiaofan-luan** [reply]
1. when node crash, milvus will replay wal in kafka to make sure data not lost
2. 1024 dimension is roughly 2K, if scalar not too large, usually each segment can store 200000 rows. When you have large data set, 2G segment size could be a reasonable size

**Reply by yhmo** [reply]
> 1. If I set the size of the segment to be relatively large, and the Growing Segment is not yet persisted in the buffer, what happens if Milvus experiences a failure and needs to restart. Is there a risk of data loss at this point?

All insert/delete/upsert requests are stored into a message queue before consuming by milvus datanode/querynode. The client sends an insert request, the insert() api will wait until the request is stored into the message queue. The the request is asynchronously processed by datanode/querynode on the server side later.
For Milvus cluster, the message queue is the Plusar service. For Milvus standalone, the message queue is a local built-in RocksDb.
If the milvus server is down unexpectedly and restarted, the datanode/querynode will retrieve the un-processed requests from the Pulsar or RocksDB to rebuild the growing segment, so the data is always safe.

> 2. hh approximately how many rows of vectors (e.g., 1024 dimensions) can be stored in 512MB?

Each dimension of a vector is stored as a float32 value, so each dimension occupies 4 bytes. 1024 dimensions will occupy 4KB. So, 512MB contains 512MB/4KB=131072 rows of vectors.

**Reply by Yang1032** [reply]
wow that'is great, thanks so much for the answers guys!

**Reply by hjlee9182** [reply]
@xiaofan-luan 
Hello, You said huge dataset, what size do you usually see as huge dataset?
